### PR TITLE
Add one more thread to `DefaultConnectionPool` for handling timeouts of async tasks

### PR DESCRIPTION
**Pull request by stIncMale**
JAVA-4146

### CODE REVIEW

**Comment by stIncMale** [comment]
Removed reviewers until I fix the OOM problem. Also, @BorisDog proposed a way to potentially simplify the solution.

**Comment by stIncMale** [comment]
Re-added reviewers.

The OOM problem is going to be solved in JAVA-4178.

The approach proposed by Boris (the same we use to implement async timeout in `NettyStream`) does not make things simpler because with `DefaultConnectionPool` we have concurrent async requests queued, and we need to remove them from the queue when they time out.

**Review by rozza — COMMENTED** [review]
Looking good - one question is merging this ticket blocked by JAVA-4178? 

I'm unclear as to OOM issues caused by this change.

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java
Comment (rozza) [inline driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:1305]: Nit: Logging seems inconsistent.  Suggest either remove the logging or improve the logging message and include it elsewhere if needed.

**Review by rozza — CHANGES_REQUESTED** [review]

**Comment by stIncMale** [comment]
> one question is merging this ticket blocked by JAVA-4178?

Right, I will merge this one after merging JAVA-4178.

> I'm unclear as to OOM issues caused by this change.

The size of `DefaultConnectionPool` was drastically increased in this PR, because we added one more thread. Not only this affects the retained Java heap size of its instances, but also increases the native heap usage for the process (because for each new Java thread a new OS thread is created). This was enough for a different problem (JAVA-4178) to start manifesting itself.

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:1202]: Given all the new synchronization logic in this class, should we add a stress/concurrency unit test to detect subtle bugs or race conditions?

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:1242]: Is this line of code outside of the lock for correctness, performance, or style? If style, I will note that I prefer to avoid early returns and would move this into an else block inside the try.  But it's personal preference so I don't insist on it.

Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:1242]: This is done for correctness: we should avoid completing callbacks while holding locks. It still may happen because a caller holds a lock while calling `DefaultConnectionPool.get`, but the class does the best it can.

File: driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java:1202]: While thinking about execution in which the timeout handling was broken, I realized that things can (and must) be much better that they currently are. I changed this PR to the draft status and opened https://github.com/mongodb/mongo-java-driver/pull/722. The commit messages there explain everything, I am also happy to discuss the matter over a call if that would be simpler.

**Pull request closed without merging**
